File: src/proofatlas/core/proof.py
"""Proof representation with history of states and applied rules."""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import List, Optional, Any, Dict, Tuple

from .state import ProofState
from proofatlas.core.fol.logic import Clause


class Rule(ABC):
    """Abstract base class for inference rules."""
    
    @property
    @abstractmethod
    def name(self) -> str:
        """Name of the rule."""
        pass
    
    @abstractmethod
    def apply(self, proof: 'Proof', *args) -> 'Proof':
        """Apply the rule to create a new proof state."""
        pass


@dataclass
class ProofStep:
    """A single step in a proof."""
    state: ProofState
    rule: Optional[Rule] = None
    selected_clause: Optional[int] = None
    generated_clauses: List[Clause] = field(default_factory=list)
    metadata: Dict[str, Any] = field(default_factory=dict)


class Proof:
    """A proof consisting of a list of ProofStates and applied rules."""
    
    def __init__(self, initial_state: Optional[ProofState] = None):
        if initial_state is None:
            initial_state = ProofState([], [])
        self.steps: List[ProofStep] = [ProofStep(initial_state)]
    
    @property
    def current_state(self) -> ProofState:
        """Get the current (latest) proof state."""
        return self.steps[-1].state
    
    @property
    def initial_state(self) -> ProofState:
        """Get the initial proof state."""
        return self.steps[0].state
    
    def apply_rule(self, rule: Rule, *args) -> 'Proof':
        """Apply a rule to create a new proof with an additional step."""
        # Rules implement their own logic and call add_step
        return rule.apply(self, *args)
    
    def add_step(self, state: ProofState, rule: Optional[Rule] = None, 
                 selected_clause: Optional[int] = None, 
                 generated_clauses: Optional[List[Clause]] = None,
                 **metadata) -> 'Proof':
        """Add a new step to the proof (used by rules)."""
        if generated_clauses is None:
            generated_clauses = []
        
        step = ProofStep(
            state=state,
            rule=rule,
            selected_clause=selected_clause,
            generated_clauses=generated_clauses,
            metadata=metadata
        )
        self.steps.append(step)
        return self
    
    @property
    def length(self) -> int:
        """Number of steps in the proof (excluding initial state)."""
        return len(self.steps) - 1
    
    @property
    def is_complete(self) -> bool:
        """Check if proof is complete (found contradiction)."""
        current = self.current_state
        # Check for empty clause (contradiction)
        for clause in current.all_clauses:
            if len(clause.literals) == 0:
                return True
        return False
    
    @property
    def is_saturated(self) -> bool:
        """Check if proof search is saturated (no unprocessed clauses)."""
        return len(self.current_state.unprocessed) == 0
    
    def get_selected_clauses(self) -> List[int]:
        """Get list of selected clause indices throughout the proof."""
        return [step.selected_clause for step in self.steps[1:] 
                if step.selected_clause is not None]
    
    def get_applied_rules(self) -> List[Rule]:
        """Get list of rules applied throughout the proof."""
        return [step.rule for step in self.steps[1:] 
                if step.rule is not None]

File: rust/src/proofs/proof.rs
//! Proof representation with history of states and selected clauses

use std::collections::HashMap;
use serde::{Serialize, Deserialize};
use serde_json::Value;

use crate::core::logic::Clause;
use super::state::ProofState;

/// Result of applying an inference rule
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RuleApplication {
    pub rule_name: String,
    pub parents: Vec<usize>,
    pub generated_clauses: Vec<Clause>,
    pub deleted_clause_indices: Vec<usize>,
    pub metadata: HashMap<String, Value>,
}

impl RuleApplication {
    /// Create a new rule application
    pub fn new(rule_name: String, parents: Vec<usize>) -> Self {
        RuleApplication {
            rule_name,
            parents,
            generated_clauses: Vec::new(),
            deleted_clause_indices: Vec::new(),
            metadata: HashMap::new(),
        }
    }
    
    /// Add generated clauses
    pub fn with_generated(mut self, clauses: Vec<Clause>) -> Self {
        self.generated_clauses = clauses;
        self
    }
    
    /// Add deleted clause indices
    pub fn with_deleted(mut self, indices: Vec<usize>) -> Self {
        self.deleted_clause_indices = indices;
        self
    }
    
    /// Add metadata
    pub fn with_metadata(mut self, key: String, value: Value) -> Self {
        self.metadata.insert(key, value);
        self
    }
}

/// A single step in a proof
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ProofStep {
    pub state: ProofState,
    pub selected_clause: Option<usize>,
    pub applied_rules: Vec<RuleApplication>,
    pub metadata: HashMap<String, Value>,
}

impl ProofStep {
    /// Create a new proof step
    pub fn new(state: ProofState) -> Self {
        ProofStep {
            state,
            selected_clause: None,
            applied_rules: Vec::new(),
            metadata: HashMap::new(),
        }
    }
    
    /// Set the selected clause
    pub fn with_selected_clause(mut self, index: usize) -> Self {
        self.selected_clause = Some(index);
        self
    }
    
    /// Add applied rules
    pub fn with_applied_rules(mut self, rules: Vec<RuleApplication>) -> Self {
        self.applied_rules = rules;
        self
    }
    
    /// Add metadata
    pub fn with_metadata(mut self, key: String, value: Value) -> Self {
        self.metadata.insert(key, value);
        self
    }
}

/// A proof consisting of a sequence of proof steps
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Proof {
    pub steps: Vec<ProofStep>,
}

impl Proof {
    /// Create a new proof with an initial state
    pub fn new(initial_state: ProofState) -> Self {
        Proof {
            steps: vec![ProofStep::new(initial_state)],
        }
    }
    
    /// Create an empty proof
    pub fn empty() -> Self {
        Proof::new(ProofState::empty())
    }
    
    /// Get the initial proof state
    pub fn initial_state(&self) -> Option<&ProofState> {
        self.steps.first().map(|step| &step.state)
    }
    
    /// Get the final proof state
    pub fn final_state(&self) -> Option<&ProofState> {
        self.steps.last().map(|step| &step.state)
    }
    
    /// Get the length of the proof (number of inference steps)
    pub fn length(&self) -> usize {
        if self.steps.is_empty() {
            return 0;
        }
        
        // If last step has no selection, don't count it as an inference step
        if self.steps.last().unwrap().selected_clause.is_none() {
            self.steps.len().saturating_sub(1)
        } else {
            self.steps.len()
        }
    }
    
    /// Add a new step to the proof
    pub fn add_step(&mut self, step: ProofStep) {
        // If the last step has no selected clause, replace it
        if let Some(last) = self.steps.last() {
            if last.selected_clause.is_none() {
                self.steps.pop();
            }
        }
        
        self.steps.push(step);
    }
    
    /// Finalize the proof by ensuring the last step has no selection
    pub fn finalize(&mut self, final_state: ProofState) {
        // Remove any existing final step with no selection
        if let Some(last) = self.steps.last() {
            if last.selected_clause.is_none() {
                self.steps.pop();
            }
        }
        
        // Add final step
        self.steps.push(ProofStep::new(final_state));
    }
    
    /// Get a specific step by index
    pub fn get_step(&self, index: usize) -> Option<&ProofStep> {
        self.steps.get(index)
    }
    
    /// Get all selected clauses in order
    pub fn get_selected_clauses(&self) -> Vec<usize> {
        self.steps.iter()
            .filter_map(|step| step.selected_clause)
            .collect()
    }
    
    /// Check if the proof found a contradiction
    pub fn found_contradiction(&self) -> bool {
        self.final_state()
            .map(|state| state.contains_empty_clause())
            .unwrap_or(false)
    }
    
    /// Get metadata history for a specific key
    pub fn get_metadata_history(&self, key: &str) -> Vec<&Value> {
        self.steps.iter()
            .filter_map(|step| step.metadata.get(key))
            .collect()
    }
}

// Include tests
#[cfg(test)]
#[path = "proof_tests.rs"]
mod tests;